src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3521,98 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    // Looks up a definitions (and paths) based on a user-supplied predicate
+    // This method should be used to generate suggestions when name resolution
+    // fails.
+    fn lookup_suggestions<FilterFn>(&mut self,
+                                    lookup_name: Name,
+                                    lookup_type: LookupTarget,
+                                    namespace: Namespace,
+                                    filter_fn: FilterFn) -> Vec<Path>
+        where FilterFn: Fn(Def) -> bool {
```

Comment:
Right now, this will report every possible path for each candidate item.
For example, consider
```rust
mod foo {
    pub mod bar {
        pub trait T {}
    }
    pub use self::bar::T;
}

mod baz {
    pub use foo;
}
```
Searching for the trait `T` would give paths `foo::T`, `foo::bar::T`, `baz::foo::T`, and `baz::foo::bar::T`.
I'm not sure if this was your intention, but I think it would be better to report one of the paths right now. It would be easiest to just report the canonical path (i.e. `foo::bar::T`).